Cross compiled binary names should changed if envs are explicitly set

Hi,

Based on the discussion #337, currently if we build with the following commands:

```bash
env GOOS=darwin GOARCH=amd64 gb build
env GOOS=linux GOARCH=amd64 gb build
env GOOS=windows GOARCH=amd64 gb build
```

it creates binaries with names:

```bash
-rwxr-xr-x   1 fatih  staff  16432868 Sep  8 00:47 images*
-rwxr-xr-x   1 fatih  staff  16528840 Sep  8 00:47 images-linux-amd64*
-rwxr-xr-x   1 fatih  staff  16358400 Sep  8 00:47 images-windows-amd64.exe*
```

Even though I've explicitly set the environment variables with `GOOS=darwin GOARCH=amd64`, the resulting binary is set to `images`. I think it could (should?) be named as `images-darwin-amd64` if the envs are set **explicitly**. 

Because I think iit would make it much more easier to group and create new artifacts (such as zip,tar, etc.. archives) based on the names. So we could create scripts such like `cp images-$GOOS-$GOARCH  out/$GOOS/` etc ...

I've written three scenarios as follows:

```Diff
diff --git a/cmd/gb/gb_test.go b/cmd/gb/gb_test.go
index 3965185..b52bf23 100644
--- a/cmd/gb/gb_test.go
+++ b/cmd/gb/gb_test.go
@@ -1501,3 +1501,39 @@ package doctest
        gb.run("doc", "doctest")
        gb.grepStdout("Package doctest tests gb doc$", "expected Package doctest tests gb doc")
 }
+
+func TestIssue346(t *testing.T) {
+       gb := T{T: t}
+       defer gb.cleanup()
+       gb.tempDir("src/p")
+       gb.tempFile("src/p/main.go", `package main
+func main() { println("hello world") }
+`)
+       gb.cd(gb.tempdir)
+       tmpdir := gb.tempDir("tmp")
+
+       gb.setenv("TMP", tmpdir)
+
+       // scenario 1: GOOS/GOARCH not set
+       name := "p"
+       gb.unsetenv("GOOS")
+       gb.unsetenv("GOARCH")
+       gb.run("build")
+       gb.mustExist(gb.path("bin", name))
+       gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p")
+
+       // scenario 2: GOOS/GOARCH are both set
+       goos := runtime.GOOS
+       name = fmt.Sprintf("p-%s-%s", goos, runtime.GOARCH)
+       gb.setenv("GOOS", goos)
+       gb.setenv("GOARCH", runtime.GOARCH)
+       gb.run("build")
+       gb.mustExist(gb.path("bin", name))
+       gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p-$GOOS-$GOARCH")
+
+       // scenario 3: just GOOS is set
+       os.Remove(gb.path("bin", name))
+       gb.unsetenv("GOARCH")
+       gb.run("build")
+       gb.mustNotExist(gb.path("bin", name))
+}
```

These are the test results for the proposed changes in ```package.go```:
```
go test -v -run 'Issue346' cmd/gb/gb_test.go                                                                                                 
=== RUN   TestIssue346
--- PASS: TestIssue346 (0.45s)
	gb_test.go:179: running /tmp/testgb612411546/testgb [build]
	gb_test.go:189: standard output:
	gb_test.go:190: p
		
	gb_test.go:179: running /tmp/testgb612411546/testgb [build]
	gb_test.go:189: standard output:
	gb_test.go:190: p
		
	gb_test.go:179: running /tmp/testgb612411546/testgb [build]
	gb_test.go:189: standard output:
	gb_test.go:190: p
		
PASS
ok  	command-line-arguments	2.550s
```

And this proves that it fails with the original implementation of ```package.go```:

```
go test -v -run 'Issue346' cmd/gb/gb_test.go                                                                                                                                                                   
=== RUN   TestIssue346
--- FAIL: TestIssue346 (0.18s)
	gb_test.go:179: running /tmp/testgb931453791/testgb [build]
	gb_test.go:189: standard output:
	gb_test.go:190: p
		
	gb_test.go:179: running /tmp/testgb931453791/testgb [build]
	gb_test.go:426: /tmp/testgb776087346/bin/p-linux-amd64 does not exist (stat /tmp/testgb776087346/bin/p-linux-amd64: no such file or directory)
FAIL
exit status 1
FAIL	command-line-arguments	2.225
```

Is that enough to open the PR?

SGTM.

On Fri, Jan 8, 2016 at 4:04 AM, Rafael Martins <notifications@github.com>
wrote:

> I've written three scenarios as follows:
>
> diff --git a/cmd/gb/gb_test.go b/cmd/gb/gb_test.go
> index 3965185..b52bf23 100644--- a/cmd/gb/gb_test.go+++ b/cmd/gb/gb_test.go@@ -1501,3 +1501,39 @@ package doctest
>         gb.run("doc", "doctest")
>         gb.grepStdout("Package doctest tests gb doc$", "expected Package doctest tests gb doc")
>  }++func TestIssue346(t *testing.T) {+       gb := T{T: t}+       defer gb.cleanup()+       gb.tempDir("src/p")+       gb.tempFile("src/p/main.go", `package main+func main() { println("hello world") }+`)+       gb.cd(gb.tempdir)+       tmpdir := gb.tempDir("tmp")++       gb.setenv("TMP", tmpdir)++       // scenario 1: GOOS/GOARCH not set+       name := "p"+       gb.unsetenv("GOOS")+       gb.unsetenv("GOARCH")+       gb.run("build")+       gb.mustExist(gb.path("bin", name))+       gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p")++       // scenario 2: GOOS/GOARCH are both set+       goos := runtime.GOOS+       name = fmt.Sprintf("p-%s-%s", goos, runtime.GOARCH)+       gb.setenv("GOOS", goos)+       gb.setenv("GOARCH", runtime.GOARCH)+       gb.run("build")+       gb.mustExist(gb.path("bin", name))+       gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p-$GOOS-$GOARCH")++       // scenario 3: just GOOS is set+       os.Remove(gb.path("bin", name))+       gb.unsetenv("GOARCH")+       gb.run("build")+       gb.mustNotExist(gb.path("bin", name))+}
>
> These are the test results for the proposed changes in package.go:
>
> go test -v -run 'Issue346' cmd/gb/gb_test.go
> === RUN   TestIssue346
> --- PASS: TestIssue346 (0.45s)
>     gb_test.go:179: running /tmp/testgb612411546/testgb [build]
>     gb_test.go:189: standard output:
>     gb_test.go:190: p
>
>     gb_test.go:179: running /tmp/testgb612411546/testgb [build]
>     gb_test.go:189: standard output:
>     gb_test.go:190: p
>
>     gb_test.go:179: running /tmp/testgb612411546/testgb [build]
>     gb_test.go:189: standard output:
>     gb_test.go:190: p
>
> PASS
> ok      command-line-arguments  2.550s
>
> And this proves that it fails with the original implementation of
> package.go:
>
> go test -v -run 'Issue346' cmd/gb/gb_test.go
> === RUN   TestIssue346
> --- FAIL: TestIssue346 (0.18s)
>     gb_test.go:179: running /tmp/testgb931453791/testgb [build]
>     gb_test.go:189: standard output:
>     gb_test.go:190: p
>
>     gb_test.go:179: running /tmp/testgb931453791/testgb [build]
>     gb_test.go:426: /tmp/testgb776087346/bin/p-linux-amd64 does not exist (stat /tmp/testgb776087346/bin/p-linux-amd64: no such file or directory)
> FAIL
> exit status 1
> FAIL    command-line-arguments  2.225
>
> Is that enough to open the PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/constabulary/gb/issues/346#issuecomment-169775489>.
>

--- a/cmd/gb/gb_test.go
+++ b/cmd/gb/gb_test.go
@@ -1501,3 +1501,39 @@ package doctest
 	gb.run("doc", "doctest")
 	gb.grepStdout("Package doctest tests gb doc$", "expected Package doctest tests gb doc")
 }
+
+func TestIssue346(t *testing.T) {
+	gb := T{T: t}
+	defer gb.cleanup()
+	gb.tempDir("src/p")
+	gb.tempFile("src/p/main.go", `package main
+func main() { println("hello world") }
+`)
+	gb.cd(gb.tempdir)
+	tmpdir := gb.tempDir("tmp")
+
+	gb.setenv("TMP", tmpdir)
+
+	// scenario 1: GOOS/GOARCH not set
+	name := "p"
+	gb.unsetenv("GOOS")
+	gb.unsetenv("GOARCH")
+	gb.run("build")
+	gb.mustExist(gb.path("bin", name))
+	gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p")
+
+	// scenario 2: GOOS/GOARCH are both set
+	goos := runtime.GOOS
+	name = fmt.Sprintf("p-%s-%s", goos, runtime.GOARCH)
+	gb.setenv("GOOS", goos)
+	gb.setenv("GOARCH", runtime.GOARCH)
+	gb.run("build")
+	gb.mustExist(gb.path("bin", name))
+	gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p-$GOOS-$GOARCH")
+
+	// scenario 3: just GOOS is set
+	os.Remove(gb.path("bin", name))
+	gb.unsetenv("GOARCH")
+	gb.run("build")
+	gb.mustNotExist(gb.path("bin", name))
+}

--- a/cmd/gb/gb_test.go
+++ b/cmd/gb/gb_test.go
@@ -1511,24 +1511,28 @@ func main() { println("hello world") }
 `)
 	gb.cd(gb.tempdir)
 	tmpdir := gb.tempDir("tmp")
-
 	gb.setenv("TMP", tmpdir)
+
+	goos := runtime.GOOS
 
 	// scenario 1: GOOS/GOARCH not set
 	name := "p"
+	if goos == "windows" {
+		name += ".exe"
+	}
 	gb.unsetenv("GOOS")
 	gb.unsetenv("GOARCH")
 	gb.run("build")
-	gb.mustExist(gb.path("bin", name))
 	gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p")
 
 	// scenario 2: GOOS/GOARCH are both set
-	goos := runtime.GOOS
 	name = fmt.Sprintf("p-%s-%s", goos, runtime.GOARCH)
+	if goos == "windows" {
+		name += ".exe"
+	}
 	gb.setenv("GOOS", goos)
 	gb.setenv("GOARCH", runtime.GOARCH)
 	gb.run("build")
-	gb.mustExist(gb.path("bin", name))
 	gb.wantExecutable(gb.path("bin", name), "expected $PROJECT/bin/p-$GOOS-$GOARCH")
 
 	// scenario 3: just GOOS is set

--- a/package.go
+++ b/package.go
@@ -2,6 +2,7 @@ package gb
 
 import (
 	"fmt"
+	"os"
 	"path/filepath"
 	"strings"
 
@@ -81,8 +82,8 @@ func (pkg *Package) Binfile() string {
 		target = filepath.Join(pkg.Workdir(), filepath.FromSlash(pkg.ImportPath), "_test", binname(pkg))
 	}
 
-	// if this is a cross compile or there are build tags, add ctxString.
-	if pkg.isCrossCompile() {
+	// if this is a cross compile or GOOS/GOARCH are both defined or there are build tags, add ctxString.
+	if pkg.isCrossCompile() || (len(os.Getenv("GOOS")) > 0 && len(os.Getenv("GOARCH")) > 0) {
 		target += "-" + pkg.ctxString()
 	} else if len(pkg.buildtags) > 0 {
 		target += "-" + strings.Join(pkg.buildtags, "-")

--- a/package.go
+++ b/package.go
@@ -83,7 +83,7 @@ func (pkg *Package) Binfile() string {
 	}
 
 	// if this is a cross compile or GOOS/GOARCH are both defined or there are build tags, add ctxString.
-	if pkg.isCrossCompile() || (len(os.Getenv("GOOS")) > 0 && len(os.Getenv("GOARCH")) > 0) {
+	if pkg.isCrossCompile() || (os.Getenv("GOOS") != "" && os.Getenv("GOARCH") != "") {
 		target += "-" + pkg.ctxString()
 	} else if len(pkg.buildtags) > 0 {
 		target += "-" + strings.Join(pkg.buildtags, "-")

